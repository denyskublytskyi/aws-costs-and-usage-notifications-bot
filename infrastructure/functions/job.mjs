import {
  CostExplorerClient,
  GetCostAndUsageCommand,
} from "@aws-sdk/client-cost-explorer";
import ms from "ms";
import compose from "lodash/fp/compose.js";
import fpFlatMapDepth from "lodash/fp/flatMapDepth.js";
import fpGroupBy from "lodash/fp/groupBy.js";
import fpMap from "lodash/fp/map.js";
import fpSumBy from "lodash/fp/sumBy.js";
import fpEntries from "lodash/fp/entries.js";
import fpOrderBy from "lodash/fp/orderBy.js";

import eventLogger from "./lib/eventLogger.mjs";
import formatDate from "./lib/formatDate.mjs";
// import exampleJSON from "./example.json" assert {type: "json"};
// import exampleJSON from "./example.json";
import sendNotification from "./lib/sendNotification.mjs";

const client = new CostExplorerClient();

const getCostAndUsage = async ({ startDate, endDate, granularity }) => {
  // return exampleJSON;
  const command = new GetCostAndUsageCommand({
    TimePeriod: {
      Start: formatDate(startDate),
      End: formatDate(endDate),
    },
    Granularity: granularity,
    Metrics: ["UnblendedCost", "UsageQuantity"],
    // Filter: {
    //   Tags: {
    //     Key: "Application",
    //     Values: ["uasystem-design"],
    //   },
    // },
    GroupBy: [
      {
        Type: "DIMENSION",
        Key: "SERVICE",
      },
    ],
  });
  return client.send(command);
};

export const prepareDataByServices = (data) => {
  return compose(
    fpOrderBy(["cost"], ["desc"]),
    fpMap(([service, entries]) => ({
      service,
      cost: fpSumBy("cost", entries),
      usage: fpSumBy("usage", entries),
    })),
    fpEntries,
    fpGroupBy("service"),
    fpFlatMapDepth(
      ({ Groups }) =>
        Groups.map(({ Keys, Metrics }) => ({
          service: Keys[0],
          cost: Math.abs(+Metrics.UnblendedCost.Amount),
          usage: Math.abs(+Metrics.UsageQuantity.Amount),
        })),
      1,
    ),
  )(data);
};

const getMessage = ({ data, title }) => {
  const totalCost = fpSumBy("cost", data);
  const byServicesMessage = data.map(
    ({ service, cost, usage }) =>
      `*${service}*\nCost: ${cost.toFixed(10)}$\nUsage: ${usage.toFixed(10)}`,
  );

  return [
    `**${title}** cost: ${totalCost.toFixed(10)}$`,
    ...byServicesMessage,
  ].join("\n\n");
};

const eventHandler = async ({ ScheduledTime }) => {
  const today = new Date(ScheduledTime);
  const yesterday = new Date(today - ms("1d"));
  const startOfTheMonth = new Date(today);
  startOfTheMonth.setDate(1);

  const response = await getCostAndUsage({
    startDate: startOfTheMonth,
    endDate: today,
    granularity: "DAILY",
  });

  const monthlyData = prepareDataByServices(response.ResultsByTime);
  const yesterdayData = prepareDataByServices(response.ResultsByTime.slice(-1));

  const monthTitle = today.toLocaleString("default", { month: "long" });
  const yesterdayTitle = yesterday.toLocaleDateString("en-GB", {
    month: "long",
    day: "numeric",
  });

  await Promise.all([
    sendNotification(
      getMessage({ data: yesterdayData, title: yesterdayTitle }),
    ),
    sendNotification(getMessage({ data: monthlyData, title: monthTitle })),
  ]);
};

export const handler = eventLogger(eventHandler);
