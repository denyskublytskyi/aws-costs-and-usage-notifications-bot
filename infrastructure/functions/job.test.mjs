import { handler, prepareDataByServices } from "./job.mjs";

import exampleJSON from "./example.json";

describe("job", () => {
  test("return Cost Explorer results", async () => {
    const result = await handler({
      ScheduledTime: new Date().toISOString(),
    });

    expect(result).toBeUndefined();
  });

  test("prepareDataByServices", async () => {
    expect(prepareDataByServices(exampleJSON.ResultsByTime)).toMatchSnapshot();
  });
});
