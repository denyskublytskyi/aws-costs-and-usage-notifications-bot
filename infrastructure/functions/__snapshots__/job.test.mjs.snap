// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`job prepareDataByServices 1`] = `
[
  {
    "cost": 0.0002699838,
    "service": "AWS Data Transfer",
    "usage": 0,
  },
  {
    "cost": 0.00026998340000000003,
    "service": "AWS Lambda",
    "usage": 37242.93113643781,
  },
  {
    "cost": 9.6e-8,
    "service": "Amazon EC2 Container Registry (ECR)",
    "usage": 0.029259878399999998,
  },
  {
    "cost": 1.3399999999999998e-8,
    "service": "Amazon DynamoDB",
    "usage": 798.0130995859,
  },
  {
    "cost": 5e-10,
    "service": "Amazon Simple Storage Service",
    "usage": 189.22589179750003,
  },
  {
    "cost": 0,
    "service": "AWS Key Management Service",
    "usage": 427,
  },
  {
    "cost": 0,
    "service": "Amazon API Gateway",
    "usage": 1423.0002361097,
  },
  {
    "cost": 0,
    "service": "Amazon Simple Queue Service",
    "usage": 76664.0000040512,
  },
  {
    "cost": 0,
    "service": "AmazonCloudWatch",
    "usage": 0.2721100088,
  },
  {
    "cost": 0,
    "service": "Tax",
    "usage": 0,
  },
  {
    "cost": 0,
    "service": "Amazon Simple Notification Service",
    "usage": 24.0000005364,
  },
  {
    "cost": 0,
    "service": "CloudWatch Events",
    "usage": 4,
  },
  {
    "cost": 0,
    "service": "AWS Cost Explorer",
    "usage": 6,
  },
]
`;
